File: core/src/lexer.rs
#![allow(unused)]
use crate::token::{Token, TokenKind};
use core::iter::Peekable;
use core::str::Chars;

pub struct Lexer<'a> {
    src: &'a str,
    chars: Peekable<Chars<'a>>,
    current_position: usize,
    start_position: usize,

    // Loom 新增：缩进控制
    indent_stack: Vec<usize>,   // 存储每一层的缩进空格数，初始为 [0]
    pending_tokens: Vec<Token>, // 用于存储 Dedent 时一次性弹出的多个 Token
}

impl<'a> Lexer<'a> {
    pub fn new(src: &'a str) -> Self {
        Self {
            src,
            chars: src.chars().peekable(),
            current_position: 0,
            start_position: 0,
            indent_stack: vec![0], // 默认顶层缩进为 0
            pending_tokens: Vec::new(),
        }
    }

    fn make_token(&self, kind: TokenKind) -> Token {
        Token::new(kind, self.start_position, self.current_position)
    }

    // 创建一个特定位置的 Token (用于 Dedent 等不由当前字符生成的 Token)
    fn make_synthetic_token(&self, kind: TokenKind, start: usize, end: usize) -> Token {
        Token::new(kind, start, end)
    }

    fn advance(&mut self) -> Option<char> {
        let c = self.chars.next()?;
        self.current_position += c.len_utf8();
        Some(c)
    }

    fn match_char(&mut self, expected: char) -> bool {
        if let Some(&c) = self.chars.peek() {
            if c == expected {
                self.advance();
                return true;
            }
        }
        false
    }

    fn peek(&mut self) -> Option<char> {
        self.chars.peek().copied()
    }
}

impl<'a> Lexer<'a> {
    pub fn next_token(&mut self) -> Token {
        // 1. 优先返回队列中的 Token (用于处理连续 Dedent)
        if let Some(token) = self.pending_tokens.pop() {
            return token;
        }

        // 2. 跳过普通空格（但不跳过换行！）
        self.skip_horizontal_whitespace();

        self.start_position = self.current_position;

        let c = match self.peek() {
            Some(c) => c,
            None => {
                // EOF 处理：如果还有缩进，需要自动补 Dedent
                if self.indent_stack.len() > 1 {
                    self.indent_stack.pop();
                    return self.make_token(TokenKind::Dedent);
                }
                return self.make_token(TokenKind::EOF);
            }
        };

        // 3. 处理换行与缩进逻辑
        if c == '\n' {
            self.advance(); // 吃掉换行
            return self.handle_newline();
        }

        if c == '#' {
            self.skip_comment_line();
            // 注释也是一种“空白”，跳过它是为了读取下一行的缩进
            return self.next_token();
        }

        // 正常 Token 解析
        self.advance(); // 消耗当前字符

        match c {
            // 标识符 (Identifier) & 下划线 (_)
            c if is_ident_start(c) => self.scan_identifier(),

            // 数字 (Number)
            c if c.is_ascii_digit() => self.scan_number(),

            // 字符串
            '"' => self.scan_string(),
            // 字符 (Loom 好像暂时没定义 char 类型，不过保留也行)
            '\'' => self.scan_char(),

            // --- 符号 ---
            '(' => self.make_token(TokenKind::LeftParen),
            ')' => self.make_token(TokenKind::RightParen),
            '{' => self.make_token(TokenKind::LeftBrace),
            '}' => self.make_token(TokenKind::RightBrace),
            '[' => self.make_token(TokenKind::LeftBracket),
            ']' => self.make_token(TokenKind::RightBracket),
            ',' => self.make_token(TokenKind::Comma),
            ':' => self.make_token(TokenKind::Colon),

            // Dot (.)
            '.' => {
                if self.match_char('.') {
                    // case: .. (Range)
                    self.make_token(TokenKind::DotDot)
                } else {
                    // case: . (Access)
                    self.make_token(TokenKind::Dot)
                }
            }

            // Plus (+)
            '+' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::PlusAssign)
                } else {
                    self.make_token(TokenKind::Plus)
                }
            }

            // Minus (-)
            '-' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::MinusAssign)
                } else {
                    self.make_token(TokenKind::Minus)
                }
            }

            // Star (*)
            '*' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::StarAssign)
                } else {
                    self.make_token(TokenKind::Star)
                }
            }

            // Slash (/)
            '/' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::SlashAssign)
                } else {
                    self.make_token(TokenKind::Slash)
                }
            }

            // Mod (%)
            '%' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::PercentAssign)
                } else {
                    self.make_token(TokenKind::Percent)
                }
            }

            // Equal (=)
            '=' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::Equal)
                } else if self.match_char('>') {
                    // case: => (Fat Arrow for inline function)
                    self.make_token(TokenKind::FatArrow)
                } else {
                    self.make_token(TokenKind::Assign)
                }
            }

            // Bang (!)
            '!' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::NotEqual)
                } else {
                    self.make_token(TokenKind::Bang)
                }
            }

            // Less (<)
            '<' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::LessEqual)
                } else {
                    self.make_token(TokenKind::LessThan)
                }
            }

            // Greater (>)
            '>' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::GreaterEqual)
                } else {
                    self.make_token(TokenKind::GreaterThan)
                }
            }

            _ => self.make_token(TokenKind::ERROR),
        }
    }

    // --- 缩进处理核心逻辑 ---

    fn handle_newline(&mut self) -> Token {
        let start_pos = self.current_position;

        // 计算下一行的缩进空格数
        let mut indent_spaces = 0;
        while let Some(&c) = self.chars.peek() {
            if c == ' ' {
                indent_spaces += 1;
                self.advance();
            } else if c == '\t' {
                indent_spaces += 4; // 假设 Tab = 4 空格，或者报错禁止 Tab
                self.advance();
            } else {
                break;
            }
        }

        // 如果紧接着又是换行或注释，忽略这一行（空行不影响缩进）
        if let Some(&next_c) = self.chars.peek() {
            if next_c == '\n' {
                self.advance();
                return self.handle_newline(); // 递归处理下一行
            }
            if next_c == '/' && self.peek_next_is('/') {
                self.skip_comment_line();
                // 注释行也不影响缩进
                return self.handle_newline();
            }
        }

        let current_indent = *self.indent_stack.last().unwrap();

        if indent_spaces > current_indent {
            // 缩进增加 -> Indent Token
            self.indent_stack.push(indent_spaces);
            self.make_synthetic_token(TokenKind::Indent, start_pos, self.current_position)
        } else if indent_spaces < current_indent {
            // 缩进减少 -> 可能产生多个 Dedent Token
            while let Some(&top) = self.indent_stack.last() {
                if top > indent_spaces {
                    self.indent_stack.pop();
                    // 这里我们先把 Token 存入 pending，因为 next_token 只能返回一个
                    // 注意：pending 是栈，先进后出，所以我们 push 进去，pop 出来正好顺序对
                    // 但这里全是 Dedent，顺序无所谓。
                    // 最后一个 Dedent 我们直接返回，剩下的存 pending
                    if *self.indent_stack.last().unwrap() > indent_spaces {
                        self.pending_tokens.push(self.make_synthetic_token(
                            TokenKind::Dedent,
                            start_pos,
                            self.current_position,
                        ));
                    } else if *self.indent_stack.last().unwrap() == indent_spaces {
                        // 匹配到了，返回最后一个 Dedent
                        return self.make_synthetic_token(
                            TokenKind::Dedent,
                            start_pos,
                            self.current_position,
                        );
                    } else {
                        // 缩进不对齐错误！(例如 4 -> 2 -> 3)
                        return self.make_token(TokenKind::ERROR);
                    }
                } else {
                    break;
                }
            }
            // 理论上上面会 return，如果不匹配会走 Error
            self.make_token(TokenKind::ERROR)
        } else {
            // 缩进不变 -> Newline Token
            self.make_synthetic_token(TokenKind::Newline, start_pos, self.current_position)
        }
    }

    // --- 辅助函数 ---

    fn skip_horizontal_whitespace(&mut self) {
        while let Some(&c) = self.chars.peek() {
            if c == ' ' || c == '\t' || c == '\r' {
                self.advance();
            } else {
                break;
            }
        }
    }

    fn skip_comment_line(&mut self) {
        // 假设注释是 //
        self.advance(); // /
        self.advance(); // /
        while let Some(&c) = self.chars.peek() {
            if c == '\n' {
                break;
            }
            self.advance();
        }
    }

    // --- 以下保持大部分原有逻辑，微调 ---

    fn scan_identifier(&mut self) -> Token {
        let start = self.start_position; // 修正：使用 start_position
        // 注意：首字符已经被 advance 消耗了，所以不需要在这里 advance
        // 但为了获取完整字符串，我们需要从 src 中切片

        while let Some(&c) = self.chars.peek() {
            if is_ident_continue(c) {
                self.advance();
            } else {
                break;
            }
        }

        let text = &self.src[self.start_position..self.current_position];

        // 关键字查找
        let kind = TokenKind::lookup_keyword(text).unwrap_or(TokenKind::Identifier);
        self.make_token(kind)
    }

    pub fn scan_number(&mut self) -> Token {
        self.consume_digits(10);
        // 简单处理 float
        if self.peek_is('.') && self.peek_next_is_digit() {
            self.advance();
            self.consume_digits(10);
            return self.make_token(TokenKind::Float);
        }
        self.make_token(TokenKind::Integer)
    }

    fn consume_digits(&mut self, radix: u32) {
        while let Some(&c) = self.chars.peek() {
            if c.is_digit(radix) || c == '_' {
                self.advance();
            } else {
                break;
            }
        }
    }

    fn peek_is(&mut self, expected: char) -> bool {
        self.peek() == Some(expected)
    }

    fn peek_next_is(&self, expected: char) -> bool {
        let mut lookahead = self.chars.clone();
        lookahead.next();
        lookahead.next() == Some(expected)
    }

    fn peek_next_is_digit(&self) -> bool {
        let mut lookahead = self.chars.clone();
        lookahead.next();
        if let Some(c) = lookahead.next() {
            c.is_ascii_digit()
        } else {
            false
        }
    }

    fn scan_string(&mut self) -> Token {
        while let Some(&c) = self.chars.peek() {
            match c {
                '"' => {
                    self.advance();
                    return self.make_token(TokenKind::StringLiteral);
                }
                '\\' => {
                    self.advance();
                    self.advance();
                } // 转义
                '\n' => return self.make_token(TokenKind::ERROR), // 禁止跨行字符串
                _ => {
                    self.advance();
                }
            }
        }
        self.make_token(TokenKind::ERROR)
    }

    fn scan_char(&mut self) -> Token {
        // 简单实现
        if let Some(c) = self.advance() {
            if c == '\'' {
                return self.make_token(TokenKind::ERROR);
            } // 空 char
            if c == '\\' {
                self.advance();
            }
        }
        if self.match_char('\'') {
            self.make_token(TokenKind::CharLiteral) // 需要在 TokenKind 加这个或者复用 Integer
        } else {
            self.make_token(TokenKind::ERROR)
        }
    }
}

fn is_ident_start(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_ident_continue(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

File: compiler/lexer.zig
const std = @import("std");
const Token = @import("token.zig").Token;
const TokenType = @import("token.zig").TokenType;
const Span = @import("utils.zig").Span;

pub const Lexer = struct {
    source: []const u8,
    start: usize = 0, // 当前 Token 的起始位置
    current: usize = 0, // 扫描探针的当前位置

    pub fn init(source: []const u8) Lexer {
        return .{
            .source = source,
        };
    }

    /// 获取下一个 Token
    pub fn next(self: *Lexer) Token {
        self.skipWhitespace();

        self.start = self.current;

        const char = self.advance() orelse return self.makeToken(.Eof);

        return switch (char) {
            // 标识符或关键字 (a-z, A-Z, _)
            'a'...'z', 'A'...'Z', '_' => self.scanIdentifier(),

            // 数字 (0-9)
            '0'...'9' => self.scanNumber(),

            // 字符串
            '"' => self.scanString(),

            // 字符
            '\'' => self.scanChar(),

            // 运算符和标点
            '(' => self.makeToken(.LParen),
            ')' => self.makeToken(.RParen),
            '{' => self.makeToken(.LBrace),
            '}' => self.makeToken(.RBrace),
            '[' => self.makeToken(.LBracket),
            ']' => self.makeToken(.RBracket),
            ',' => self.makeToken(.Comma),
            ';' => self.makeToken(.Semicolon),
            ':' => self.makeToken(.Colon),
            '#' => self.makeToken(.Hash),
            '?' => self.makeToken(.Question),
            '$' => self.makeToken(.Dollar),
            '@' => self.makeToken(.At),
            // Dot 家族处理
            '.' => {
                // 1. 检查 ...
                if (self.match('.')) {
                    // 检查是否是 ... (变长参数)
                    if (self.match('.')) {
                        return self.makeToken(.Ellipsis);
                    }
                    // 检查是否是 ..= (范围匹配) [新增]
                    else if (self.match('=')) {
                        return self.makeToken(.DotDotEqual);
                    }
                    // 否则就是普通的 ..
                    return self.makeToken(.DotDot);
                }
                // 2. 检查 .?
                else if (self.match('?')) {
                    return self.makeToken(.DotQuestion);
                }
                // 3. 检查 .*
                else if (self.match('*')) {
                    return self.makeToken(.DotStar);
                }
                // 4. 检查 .<
                else if (self.match('<')) {
                    return self.makeToken(.DotLessThan);
                }
                // 5. 普通点 .
                else {
                    return self.makeToken(.Dot);
                }
            },

            '+' => self.matchAssign(.Plus, .PlusAssign),
            '-' => self.matchAssign(.Minus, .MinusAssign),
            '*' => self.matchAssign(.Star, .StarAssign),
            '%' => self.matchAssign(.Percent, .PercentAssign),
            '/' => {
                // 1. 检查是否是单行注释 //
                if (self.match('/')) {
                    self.skipCommentLine();
                    return self.next(); // 递归调用，寻找下一个有效 Token
                }
                // 2. 检查是否是多行注释 /*
                else if (self.match('*')) {
                    // 进入这里时，已经消耗了 "/*"
                    // 所以 skipCommentBlock 内部 depth 初始为 1
                    self.skipCommentBlock();
                    return self.next(); // 递归调用
                }
                // 3. 检查是否是除法赋值 /=
                else if (self.match('=')) {
                    return self.makeToken(.SlashAssign);
                }
                // 4. 普通除号 /
                else {
                    return self.makeToken(.Slash);
                }
            },

            '=' => if (self.match('=')) self.makeToken(.Equal) else if (self.match('>')) self.makeToken(.Arrow) else self.makeToken(.Assign),
            '!' => if (self.match('=')) self.makeToken(.NotEqual) else self.makeToken(.Bang),
            '<' => {
                // 检查是否是左移 <<
                if (self.match('<')) {
                    // 检查是否是左移赋值 <<=
                    if (self.match('=')) return self.makeToken(.LShiftAssign);
                    return self.makeToken(.LShift);
                }
                // 检查是否是小于等于 <=
                if (self.match('=')) return self.makeToken(.LessEqual);
                // 否则就是小于 <
                return self.makeToken(.LessThan);
            },

            '>' => {
                // 检查是否是右移 >>
                if (self.match('>')) {
                    // 检查是否是右移赋值 >>=
                    if (self.match('=')) return self.makeToken(.RShiftAssign);
                    return self.makeToken(.RShift);
                }
                // 检查是否是大于等于 >=
                if (self.match('=')) return self.makeToken(.GreaterEqual);
                // 否则就是大于 >
                return self.makeToken(.GreaterThan);
            },

            // 位运算
            '&' => self.matchAssign(.Ampersand, .AmpersandAssign),
            '|' => self.matchAssign(.Pipe, .PipeAssign),
            '^' => self.matchAssign(.Caret, .CaretAssign),
            '~' => self.makeToken(.Tilde),

            else => self.makeToken(.Illegal),
        };
    }

    // === 核心扫描逻辑 ===

    fn scanIdentifier(self: *Lexer) Token {
        while (isAlphaNumeric(self.peek())) {
            _ = self.advance();
        }

        const text = self.source[self.start..self.current];
        // 查表
        const tag = Token.keywords.get(text) orelse .Identifier;
        return self.makeToken(tag);
    }

    fn scanNumber(self: *Lexer) Token {
        // 1. 处理进制前缀 (0x, 0b, 0o)
        // 只有以 '0' 开头才可能是进制前缀
        if (self.source[self.start] == '0') {
            // start 指向 '0'，current 已经在 '0' 后面了，所以 peek 看的是第二个字符
            const next_char = self.peek();

            switch (next_char) {
                'x', 'X' => {
                    _ = self.advance(); // 吃掉 'x'
                    self.consumeDigits(16); // 扫描十六进制
                    return self.makeToken(.IntLiteral);
                },
                'b', 'B' => {
                    _ = self.advance(); // 吃掉 'b'
                    self.consumeDigits(2); // 扫描二进制
                    return self.makeToken(.IntLiteral);
                },
                'o', 'O' => {
                    _ = self.advance(); // 吃掉 'o'
                    self.consumeDigits(8); // 扫描八进制
                    return self.makeToken(.IntLiteral);
                },
                else => {
                    // 只是一个普通的 0，或者 0.xxxx，或者 0123
                    // 继续往下走，进入十进制逻辑
                },
            }
        }

        // 2. 扫描整数部分 (十进制)
        self.consumeDigits(10);

        // 3. 处理小数部分 (Float)
        // 关键逻辑：如果是 '.'，必须确认 '.' 后面紧跟着数字，才算是浮点数。
        // 否则可能是 1.method() 或者是 1..10 (Range)
        if (self.peek() == '.' and isDigit(self.peekNext())) {
            _ = self.advance(); // 吃掉 '.'
            self.consumeDigits(10); // 扫描小数部分

            // 扫描完小数后，还可以继续跟指数部分，如 1.2e10
            _ = self.tryScanExponent();
            return self.makeToken(.FloatLiteral);
        }

        // 4. 处理没有小数点的指数部分 (如 1e10)
        // 这也是浮点数
        if (self.tryScanExponent()) {
            return self.makeToken(.FloatLiteral);
        }

        // 既没有小数点，也没有指数，就是普通的整数
        return self.makeToken(.IntLiteral);
    }

    /// 尝试扫描指数部分 (e/E)，如果有返回 true
    fn tryScanExponent(self: *Lexer) bool {
        const c = self.peek();
        if (c == 'e' or c == 'E') {
            _ = self.advance(); // 吃掉 'e'

            // 指数部分可以有正负号: 1e-10, 1e+5
            const next_c = self.peek();
            if (next_c == '+' or next_c == '-') {
                _ = self.advance();
            }

            self.consumeDigits(10);
            return true;
        }
        return false;
    }

    /// 通用的数字消费函数，支持下划线分隔符
    fn consumeDigits(self: *Lexer, radix: u8) void {
        while (true) {
            const c = self.peek();
            if (c == '_') {
                _ = self.advance();
                continue;
            }

            // 根据进制判断是否是有效数字
            const is_valid = switch (radix) {
                2 => isBinDigit(c),
                8 => isOctDigit(c),
                10 => isDigit(c),
                16 => isHexDigit(c),
                else => false, // unreachable
            };

            if (is_valid) {
                _ = self.advance();
            } else {
                break;
            }
        }
    }

    fn scanString(self: *Lexer) Token {
        while (true) {
            const char = self.peek();
            switch (char) {
                0 => return self.makeToken(.Illegal), // 未闭合就结束
                '"' => {
                    _ = self.advance(); // 吞掉右引号
                    break;
                },
                '\\' => {
                    _ = self.advance(); // 跳过转义
                    _ = self.advance();
                },
                else => {
                    _ = self.advance();
                },
            }
        }
        return self.makeToken(.StringLiteral);
    }

    fn scanChar(self: *Lexer) Token {
        // 刚吃掉了左边的单引号 '，现在处于字符内容的第一个字节
        const c = self.peek();

        // 1. 处理转义字符 (以 \ 开头)
        if (c == '\\') {
            _ = self.advance(); // 吃掉反斜杠 '\'

            const escaped = self.peek();
            switch (escaped) {
                // 简单单字符转义
                'n', 'r', 't', '\\', '\'', '\"', '0' => {
                    _ = self.advance();
                },

                // 十六进制转义: \xNN
                'x' => {
                    _ = self.advance(); // 吃掉 'x'
                    // 必须严格吃掉两个十六进制位
                    if (!self.consumeHexDigits(2)) return self.makeToken(.Illegal);
                },

                // Unicode 转义: \u{...}
                'u' => {
                    _ = self.advance(); // 吃掉 'u'
                    if (self.peek() != '{') return self.makeToken(.Illegal);
                    _ = self.advance(); // 吃掉 '{'

                    // 吃掉中间的十六进制位，直到 '}'
                    // 限制最大长度防止死循环，比如 Unicode 最大 6 位 hex
                    var length: usize = 0;
                    while (isHexDigit(self.peek())) {
                        _ = self.advance();
                        length += 1;
                        if (length > 6) return self.makeToken(.Illegal);
                    }

                    if (self.peek() != '}') return self.makeToken(.Illegal);
                    _ = self.advance(); // 吃掉 '}'
                },

                else => return self.makeToken(.Illegal), // 未知的转义，比如 \q
            }
        }
        // 2. 处理普通字符 (包括 UTF-8 多字节字符)
        else if (c != '\'' and c != 0) {
            const len = std.unicode.utf8ByteSequenceLength(c) catch {
                return self.makeToken(.Illegal);
            };

            // 吃掉该字符的所有字节
            var i: usize = 0;
            while (i < len) : (i += 1) {
                _ = self.advance();
            }
        }
        // 3. 空字符 '' 或者直接遇到 EOF
        else {
            return self.makeToken(.Illegal);
        }

        // 4. 必须以单引号闭合
        if (self.match('\'')) {
            return self.makeToken(.CharLiteral);
        }

        return self.makeToken(.Illegal);
    }

    fn consumeHexDigits(self: *Lexer, count: usize) bool {
        var i: usize = 0;
        while (i < count) : (i += 1) {
            if (isHexDigit(self.peek())) {
                _ = self.advance();
            } else {
                return false;
            }
        }
        return true;
    }

    // === 辅助工具  ===

    // 前进一格并返回字符
    fn advance(self: *Lexer) ?u8 {
        if (self.current >= self.source.len) return null;
        const c = self.source[self.current];
        self.current += 1;
        return c;
    }

    // 仅查看当前字符
    fn peek(self: *Lexer) u8 {
        if (self.current >= self.source.len) return 0;
        return self.source[self.current];
    }

    // 查看下一个字符 (Lookahead 1)
    fn peekNext(self: *Lexer) u8 {
        if (self.current + 1 >= self.source.len) return 0;
        return self.source[self.current + 1];
    }

    // 匹配当前字符，如果匹配则前进
    fn match(self: *Lexer, expected: u8) bool {
        if (self.current >= self.source.len) return false;
        if (self.source[self.current] != expected) return false;
        self.current += 1;
        return true;
    }

    // 语法糖：匹配 =, 如 +=, -=
    fn matchAssign(self: *Lexer, single: TokenType, double: TokenType) Token {
        if (self.match('=')) {
            return self.makeToken(double);
        }
        return self.makeToken(single);
    }

    fn makeToken(self: *Lexer, tag: TokenType) Token {
        return .{
            .tag = tag,
            .span = Span.new(self.start, self.current),
        };
    }

    fn skipWhitespace(self: *Lexer) void {
        while (true) {
            const c = self.peek();
            switch (c) {
                ' ', '\t', '\r', '\n' => _ = self.advance(),
                else => break,
            }
        }
    }

    fn skipCommentLine(self: *Lexer) void {
        while (self.peek() != '\n' and self.peek() != 0) {
            _ = self.advance();
        }
    }

    fn skipCommentBlock(self: *Lexer) void {
        var depth: usize = 1;

        while (depth > 0) {
            const c = self.peek();

            // 1. 检查是否到达文件末尾 (EOF)
            // 如果文件结束了注释还没闭合，这是一个错误，
            // 这里直接停止，记得在Parser中检查和处理报错
            if (c == 0 and self.current >= self.source.len) {
                return;
            }

            // 2. 检查嵌套开始 /*
            if (c == '/' and self.peekNext() == '*') {
                // 吃掉 '/'
                _ = self.advance();
                // 吃掉 '*'
                _ = self.advance();
                depth += 1;
                continue;
            }

            // 3. 检查嵌套结束 */
            if (c == '*' and self.peekNext() == '/') {
                // 吃掉 '*'
                _ = self.advance();
                // 吃掉 '/'
                _ = self.advance();
                depth -= 1;
                continue;
            }

            // 4. 普通字符，跳过
            _ = self.advance();
        }
    }
};

fn isAlphaNumeric(c: u8) bool {
    return isAlpha(c) or isDigit(c);
}

fn isAlpha(c: u8) bool {
    return (c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z') or c == '_';
}

fn isDigit(c: u8) bool {
    return c >= '0' and c <= '9';
}

fn isHexDigit(c: u8) bool {
    return (c >= '0' and c <= '9') or (c >= 'a' and c <= 'f') or (c >= 'A' and c <= 'F');
}

fn isBinDigit(c: u8) bool {
    return c == '0' or c == '1';
}

fn isOctDigit(c: u8) bool {
    return c >= '0' and c <= '7';
}

// === 测试区 ===

// 核心测试辅助函数
fn expectTokens(source: []const u8, expected_tags: []const TokenType) !void {
    var lex = Lexer.init(source);

    for (expected_tags) |expected_tag| {
        const next_token = lex.next();
        // 验证 Token 类型是否匹配
        try std.testing.expectEqual(expected_tag, next_token.tag);
    }

    // 最后必须确保读到的是 EOF
    const end_token = lex.next();
    try std.testing.expectEqual(TokenType.Eof, end_token.tag);
}

test "Lexer - Basic Symbols" {
    try expectTokens("= + ( ) { }", &.{ .Assign, .Plus, .LParen, .RParen, .LBrace, .RBrace });

    try expectTokens("#len ?opt val.? val.* func.<T> ... @bitcast 1..=10", &.{
        .Hash,
        .Identifier,
        .Question,
        .Identifier,
        .Identifier,
        .DotQuestion,
        .Identifier,
        .DotStar, // val.*
        .Identifier,
        .DotLessThan,
        .Identifier,
        .GreaterThan,
        .Ellipsis,
        .At, // @
        .Identifier, // bitcast
        .IntLiteral,
        .DotDotEqual, // ..=
        .IntLiteral,
    });
}

test "Lexer - Numbers" {
    try expectTokens(
        \\123 
        \\123_456 
        \\0xDEAD_BEEF 
        \\0b1010
        \\3.14 
        \\0.5 
        \\1e10 
        \\2.5e-3
    , &.{
        .IntLiteral, // 123
        .IntLiteral, // 123_456
        .IntLiteral, // 0xDEAD_BEEF
        .IntLiteral, // 0b1010
        .FloatLiteral, // 3.14
        .FloatLiteral, // 0.5
        .FloatLiteral, // 1e10
        .FloatLiteral, // 2.5e-3
    });
}

test "Lexer - Range vs Float" {
    // 1..5 应该是 Int, DotDot, Int
    // 1.5 应该是 Float
    try expectTokens("1..5", &.{ .IntLiteral, .DotDot, .IntLiteral });
    try expectTokens("1.5", &.{.FloatLiteral});
}

test "Lexer - Chars" {
    try expectTokens(
        \\'a' '\n' '\'' '\\' '\xAF' '\u{1F600}'
    , &.{
        .CharLiteral,
        .CharLiteral,
        .CharLiteral,
        .CharLiteral,
        .CharLiteral,
        .CharLiteral,
    });
}

test "Lexer - Strings" {
    try expectTokens(
        \\"hello" "world\n"
    , &.{
        .StringLiteral,
        .StringLiteral,
    });
}

test "Lexer - Keywords and Identifiers" {
    try expectTokens(
        \\fn let mut my_var Return return static
    , &.{
        .Fn,
        .Let,
        .Mut, // mut
        .Identifier, // my_var
        .Identifier, // Return (大写)
        .Return,
        .Static, // static
    });
}

test "Lexer - Comments" {
    const code =
        \\let a = 1; // single line
        \\/* multi 
        \\   line */
        \\let b = 2;
        \\/* nested /* inside */ outside */
        \\let c = 3;
    ;

    try expectTokens(code, &.{
        .Let, .Identifier, .Assign, .IntLiteral, .Semicolon, // let a = 1;
        .Let, .Identifier, .Assign, .IntLiteral, .Semicolon, // let b = 2;
        .Let, .Identifier, .Assign, .IntLiteral, .Semicolon, // let c = 3;
    });
}

test "Lexer - Span Correctness" {
    const code = "let a";
    var lex = Lexer.init(code);

    const t1 = lex.next(); // let
    try std.testing.expectEqual(TokenType.Let, t1.tag);
    try std.testing.expectEqual(0, t1.span.start);
    try std.testing.expectEqual(3, t1.span.end);

    const t2 = lex.next(); // a
    try std.testing.expectEqual(TokenType.Identifier, t2.tag);
    try std.testing.expectEqual(4, t2.span.start); // 注意中间有个空格
    try std.testing.expectEqual(5, t2.span.end);
}

test "Lexer - Slice Content" {
    const code = "var foo = 123;";
    var lex = Lexer.init(code);

    _ = lex.next(); // var

    const t_ident = lex.next(); // foo
    // 验证 Span 切出来的字符串是不是 "foo"
    try std.testing.expect(std.mem.eql(u8, "foo", code[t_ident.span.start..t_ident.span.end]));
}
